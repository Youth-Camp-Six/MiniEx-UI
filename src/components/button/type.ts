import { ReactNode } from 'react';

type ButtonSize = 'lg' | 'sm';

type ButtonType = 'primary' | 'default' | 'danger' | 'link';

export interface BaseButtonProps {
  className?: string;
  disabled?: boolean;
  size?: ButtonSize;
  btnType?: ButtonType;
  children: ReactNode;
}

// HTML button元素的属性结合上自定义属性
type buttonProps = BaseButtonProps & React.ButtonHTMLAttributes<HTMLElement>;

// 将Button组件的所有属性转化为可选的
export type ButtonProps = Partial<buttonProps>;
