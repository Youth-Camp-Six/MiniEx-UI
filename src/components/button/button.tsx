import React, { memo } from 'react';
import classNames from 'classnames';
import { ButtonProps } from './type';

export const Button: React.FC<ButtonProps> = (props) => {
  const { className, disabled, size, btnType, children, ...restProps } = props;

  const classes = classNames('btn', className, {
    [`btn-${btnType}`]: btnType,
    [`btn-${size}`]: size,
    disabled,
  });

  return (
    <button className={classes} disabled={disabled} {...restProps}>
      {children}
    </button>
  );
};

Button.defaultProps = {
  disabled: false,
  btnType: 'default',
};

export default memo(Button);
