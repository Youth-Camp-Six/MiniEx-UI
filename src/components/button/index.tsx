import Button from './button';

export * from './type';
export { Button };
