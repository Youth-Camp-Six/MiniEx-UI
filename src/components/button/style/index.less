@import './mixin.less';

.btn {
  position: relative;
  display: inline-block;
  color: var(--mi-black-text-color);
  font-weight: 500;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  background-image: none;
  border: 1px solid transparent;
  cursor: pointer;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;

  .mixin-button-size(
    0.45rem,
    0.75rem,
    var(--mi-font-size-medium),
    var(--mi-border-radius)
  );
  // 禁用时
  &.disabled,
  &[disabled] {
    box-shadow: none;
    cursor: not-allowed;
    opacity: var(--mi-disabled-opacity);

    > * {
      // 阻止字元素鼠标事件
      pointer-events: none;
    }
  }
}

.btn-lg {
  .mixin-button-size(
    0.6rem,
    1rem,
    var(--mi-font-size-large),
    var(--mi-border-radius)
  );
}

.btn-sm {
  .mixin-button-size(
    0.35rem,
    0.5rem,
    var(--mi-font-size-small),
    var(--mi-border-radius)
  );
}

.btn-primary {
  .mixin-button-style(
    var(--mi-primary-color),
    var(--mi-primary-color),
    var(--mi-white-text-color),
    var(--mi-primary-color-hover),
    var(--mi-primary-color-hover),
    var(--mi-white-text-color),
    var(--mi-primary-color-active),
    var(--mi-primary-color-active),
    var(--mi-white-text-color),
  );
}

.btn-danger {
  .mixin-button-style(
    var(--mi-error-color),
    var(--mi-error-color),
    var(--mi-white-text-color),
    var(--mi-error-color-hover),
    var(--mi-error-color-hover),
    var(--mi-white-text-color),
    var(--mi-error-color-active),
    var(--mi-error-color-active),
    var(--mi-white-text-color),
  );
}

.btn-default {
  .mixin-button-style(
    var(--mi-normal-color);
    var(--mi-normal-color-dark);
    var(---mi-normal-text-color);
    var(--mi-normal-color-hover);
    var(--mi-normal-color-dark);
    var(---mi-normal-text-color);
    var(--mi-normal-color-active);
    var(--mi-normal-color-dark);
    var(---mi-normal-text-color);
  );
}

.btn-link {
  .mixin-button-style(
    transparent,
    none,
    var(--mi-primary-color),
    var(--mi-bgcolor-2),
    var(--mi-bgcolor-2),
    var(--mi-primary-color-hover),
    var(--mi-bgcolor-3),
    var(--mi-bgcolor-3),
    var(--mi-primary-color-active)
  );

  box-shadow: none;
}
