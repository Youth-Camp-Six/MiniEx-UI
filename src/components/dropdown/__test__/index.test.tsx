/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Dropdown, DropdownOptions } from '..';
import { Button } from '../../button';

const options: DropdownOptions[] = [
  {
    label: '西游记',
    value: '1',
  },
  {
    label: '水浒传',
    value: '2',
  },
  {
    label: '三国演义',
    value: '3',
  },
  {
    label: '红楼梦',
    value: '4',
  },
];
describe('Dropdown', () => {
  // 快照测试
  it('test case1 for ', () => {
    const wrapper = render(
      <div>
        <Dropdown options={options}>
          <Button>hover</Button>
        </Dropdown>
        <Dropdown options={options} trigger='click'>
          <Button>click</Button>
        </Dropdown>
      </div>
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });
});
