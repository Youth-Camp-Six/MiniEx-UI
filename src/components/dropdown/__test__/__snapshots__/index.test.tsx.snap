// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown test case1 for  1`] = `
<div>
  <div
    aria-expanded="false"
    class="mi-popover"
    trigger="hover"
  >
    <label>
      <div
        class="trigger"
      >
        <button
          class="btn btn-default"
        >
          hover
        </button>
      </div>
    </label>
  </div>
  <div
    aria-expanded="false"
    class="mi-popover"
    trigger="click"
  >
    <label>
      <div
        class="trigger"
      >
        <button
          class="btn btn-default"
        >
          click
        </button>
      </div>
    </label>
  </div>
</div>
`;
